tdrs-backend/tdpservice/parsers/duplicate_manager.py

```diff
@@ -0,0 +1,203 @@
+    """Container class to manage hashed values for records of the same CASE_NUMBER and RPT_MONTH_YEAR."""
+
+    def __init__(self, my_hash, manager_error_dict, generate_error):
+        self.my_hash = my_hash
+        self.manager_error_dict = manager_error_dict
+        self.generate_error = generate_error
+        self.record_ids = dict()
+        self.record_hashes = dict()
+        self.partial_hashes = dict()
+        self.error_precedence = ErrorPrecedence()
+        self.num_errors = 0
+        self.should_remove_from_db = False
+        self.current_line_number = None
+
+    def set_should_remove_from_db(self, should_remove):
+        """Set should remove from DB."""
+        self.should_remove_from_db = should_remove
+
+    def has_errors(self):
+        """Return case duplicate error state."""
+        return self.num_errors > 0
+
+    def get_records_for_post_parse_deletion(self):
+        """Return record ids if case has duplicate errors."""
+        if self.num_errors > 0 and self.should_remove_from_db:
+            return self.record_ids
+        return dict()
+
+    def __generate_error(self, err_msg, record, schema, has_precedence, is_new_max_precedence):
+        """Add an error to the managers error dictionary.
+
+        @param err_msg: string representation of the error message
+        @param record: a Django model representing a datafile record
+        @param schema: the schema from which the record was created
+        @param has_precedence: boolean indicating if this incoming error has equivalent precedence to current errors
+        @param is_new_max_precedence: boolean indicating if this error has the new max precedence
+        """
+        if has_precedence:
+            error = self.generate_error(
+                        error_category=ParserErrorCategoryChoices.CASE_CONSISTENCY,
+                        schema=schema,
+                        record=record,
+                        field=None,
+                        error_message=err_msg,
+                    )
+            if is_new_max_precedence:
+                self.manager_error_dict[self.my_hash] = [error]
+            else:
+                self.manager_error_dict.setdefault(self.my_hash, []).append(error)
+            self.num_errors = len(self.manager_error_dict[self.my_hash])
+
+    def __get_partial_hash(self, record):
+        partial_hash = None
+        if record.RecordType in {"T1", "T4"}:
+            partial_hash = hash(record.RecordType + str(record.RPT_MONTH_YEAR) + record.CASE_NUMBER)
+        elif record.RecordType in {"T2", "T3", "T5"}:
+            partial_hash = hash(record.RecordType + str(record.RPT_MONTH_YEAR) + record.CASE_NUMBER +
+                                str(record.FAMILY_AFFILIATION) + record.DATE_OF_BIRTH + record.SSN)
+        return partial_hash
+
+    def __skip_partial(self, record):
+        skip_partial = False
+        if record.RecordType == "T2":
+            skip_partial = record.FAMILY_AFFILIATION in {3, 5}
```

Comment:
i'm a little concerned that this logic is buried pretty deep within parsing/validation (both in terms of runtime logic and code/file structure) but makes direct reference to possible values for a specific field in specific schemas (some of which are sort of not captured here, either, becuase of how `RecordType` is used). I wonder if there's a better place to put this, perhaps on the schema itself? this may be better left as a future note as i could see wanting to explore some different options.

Reply:
- Because duplicate detection is program type agnostic we should be able to move this check up into the schema itself to alleviate the testing required to hit all these branches.